### DISCUSSION TITLE
`drop = remove commit` doesn't work in `git rebase`

### DISCUSSION THREAD

**Discussion by k4leg** [magit/magit#4690] [Ideas]
Before:
```git-rebase-todo
pick 0000000 msg
```

Expected behavior (press "d"):
```git-rebase-todo
drop 0000000 msg
```

Observed behavior (press "d"):
```git-rebase-todo
# pick 0000000 msg
```

How to reproduce:
```shell-session
$ mkdir q && cd q && git init
$ echo x > a && echo x > b
$ git add a && git commit -m a
$ git add b && git commit -m b
$ env VISUAL='emacsclient -c' git rebase -i HEAD~  # Drop commit.
error: nothing to do
```
And commit isn't removed (`drop` removes commit, not comment).

```
Magit 20220503.1245 [>= 3.3.0-git], Git 2.36.1, Emacs 28.1, gnu/linux
```

**Comment by tarsius** [comment]
Press `k` (kill) to do that. Magit (like many read-only Emacs modes) uses `p` for goto **p**revious line (and `n` for next line). Unless you have done `(setq git-rebase-show-instructions nil)` you should be shown the available key bindings (though as I have just noticed their display could be improved).

**Reply by k4leg** [reply]
Oh sorry, I use `evil-collections`.  But `git-rebase-kill-line` function doesn't drop (remove) commit, it just comment commit (`k` key docstring says "drop the commit at point", i. e. `drop`, not comment).

**Reply by tarsius** [reply]
It drops the commit in the sense that it tells git to so. It could do that by removing the line, but instead it does it by commenting out the line, which has the same effect. The benefit of doing it that way is that it becomes much easier to undo the dropping, by pressing `c` on a commit that is scheduled for being dropped.

**Reply by k4leg** [reply]
See my _how to reproduce_.  In this example, commit can be removed with `drop 0000000 msg`.  `# pick 0000000 msg` (i. e. call `git-rebase-kill-line` or press `k` key) gives error (`error: nothing to do`).  How can I remove commit with magit in this example?

**Reply by tarsius** [reply]
For this usecase (discarding the last commit, or N last commits) `git reset --hard` (`X h`, `magit-reset-hard`) is more suitable.

**Reply by kyleam** [reply]
I agree with the `git rebase --hard` suggestion.  You could also use `magit-rebase-remove-commit`.  

But if you do find yourself in a rebase todo buffer with all the lines commented, you can convince Git to proceed by adding a `noop` line (bound to `z` by default).

Of course, we _could_ add a command that inserts the `drop` action.  (Given the current minimum Git version, simply switching `git-rebase-kill-line` to use `drop` isn't an option `drop` isn't available until Git v2.6.)  However, I don't see much point in supporting the `drop` action given that the same end state can already be achieved in several other ways.

**Reply by mtbar131** [reply]
@tarsius I think supporting `drop` would be a good addition. I was trying to rebase and wanted to drop a commit but git failed with following error:
```
Warning: some commits may have been dropped accidentally.


Use 'git config rebase.missingCommitsCheck' to change the level of warnings.
The possible behaviours are: ignore, warn, error.
```

I can always change the git config to just log a warning but I don't want that (mainly because warnings can go unnoticed, especially when using magit). I want git to always throw an error and force me to explicitly drop a commit if I really mean to. Would it be possible to add support for `drop` during rebase?

**Reply by raszi** [reply]
`magit-reset-hard` is not the same AFAICS, because we would also lose the uncommitted changes. The good thing about rebase and drop is that Magit autostashes the changes and reappplies the stashed modifications.

It was a surprise to me that drop wasn't working in this case.

**Reply by tarsius** [reply]
Okay, I'll add it.

**Comment by tarsius** [comment]
I've added `git-rebase-drop` and bound `d` to it in 479c4670806dda3589853fa8054c227b89506be1.
